### PR TITLE
Implementing a ZIP adapter

### LINKED ISSUE #1218: ZIP Adapter
### Question

Hi.

I have ported the ZIP Adapter to Flysystem V2. Tests run fine locally. I would like to open source it but I'm not sure how. There are 3 options available for me : 

- Including it in this package
- Updating https://github.com/thephpleague/flysystem-ziparchive
- Make my own package

Which one do you prefer ? All options suit me.

### PR DESCRIPTION
**Pull request by Lctrs**
Closes #1218

### CODE REVIEW

**Comment by Lctrs** [comment]
@frankdejonge Failures are unrelated. Some PHPStan new issue.

File: src/Zip/UnableToOpenZip.php [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+        ZipArchive::ER_SEEK => 'Seek error',
+    ];
+
+    public static function atLocation(string $location, int $errorCode): self
+    {
+        return new self(sprintf(
+            'Unable to write file at location: %s. %s',
+            $location,
+            self::$errorCodesMapping[$errorCode] ?? ''
+        ));
```
Comment (OskarStark) [inline src/Zip/UnableToOpenZip.php]:
```suggestion
        return new self(trim(sprintf(
            'Unable to write file at location: %s. %s',
            $location,
            self::$errorCodesMapping[$errorCode] ?? ''
        )));
```

File: src/Zip/ZipAdapter.php [hunk 2]
```diff
@@ -0,0 +1,390 @@
...
+    /** @var string */
+    private $filename;
+    /** @var ZipArchive|null */
+    private $archive;
+    /** @var PathPrefixer */
+    private $prefixer;
+    /** @var MimeTypeDetector */
+    private $mimeTypeDetector;
+    /** @var VisibilityConverter */
+    private $visibility;
```
Comment (OskarStark) [inline src/Zip/ZipAdapter.php]: Why dont you use property types? Is it not possible due toe the min PHP version?

File: src/Zip/ZipAdapter.php [hunk 3]
```diff
@@ -0,0 +1,390 @@
...
+
+    public function __construct(
+        string $filename,
+        string $root,
+        ?MimeTypeDetector $mimeTypeDetector = null,
+        ?VisibilityConverter $visibility = null
+    ) {
```
Comment (OskarStark) [inline src/Zip/ZipAdapter.php]: Other adapter does not break constructor arguments, they use longline, but this should be decided by @frankdejonge 👍

File: src/Zip/ZipAdapter.php [hunk 4]
```diff
@@ -0,0 +1,390 @@
...
+        if ($contents === false) {
+            throw UnableToReadFile::fromLocation($path);
+        }
+
+        return $contents;
+    }
+
+    /**
+     * @inheritDoc
+     */
```
Comment (OskarStark) [inline src/Zip/ZipAdapter.php]:
```suggestion
```

File: src/Zip/ZipAdapter.php [hunk 5]
```diff
@@ -0,0 +1,390 @@
...
+        if ($visibility === null) {
+            return;
+        }
+
+        $this->setVisibility($path, $visibility);
+    }
+
+    /**
+     * @inheritDoc
+     */
```
Comment (OskarStark) [inline src/Zip/ZipAdapter.php]:
```suggestion
```

File: src/Zip/ZipAdapter.php [hunk 6]
```diff
@@ -0,0 +1,390 @@
...
+    public function createDirectory(string $path, Config $config): void
+    {
+        $this->ensureDirectoryExists(
+            $this->prefixer->prefixPath($path)
+        );
+    }
+
+    /**
+     * @inheritDoc
+     */
```
Comment (OskarStark) [inline src/Zip/ZipAdapter.php]:
```suggestion
```

same for the others

File: src/Zip/ZipAdapter.php (hunk 2 above)
Comment (Lctrs) [reply src/Zip/ZipAdapter.php]: Yes. V2 is compatible with PHP >= 7.2.

File: src/Zip/ZipAdapter.php (hunk 3 above)
Comment (Lctrs) [reply src/Zip/ZipAdapter.php]: Don't know what you mean ? At least `LocalAdapter` and `FtpAdapter` are breaking constructor arguments.

File: src/Zip/UnableToOpenZip.php (hunk 1 above)
Comment (Lctrs) [reply src/Zip/UnableToOpenZip.php]: Adressed.

File: src/Zip/ZipAdapter.php (hunk 6 above)
Comment (Lctrs) [reply src/Zip/ZipAdapter.php]: All have been adressed.

File: src/Zip/ZipAdapter.php (hunk 3 above)
Comment (frankdejonge) [reply src/Zip/ZipAdapter.php]: Breaking over multiple lines is fine 👍

**Comment by frankdejonge** [comment]
@Lctrs I've fixed those PHPStan errors on the 2.x branch.

This all looks really good btw! I've got some things that needs to be changed (certain exception types that are thrown should be caught and rethrown). How do you want to collaborate on this? I can just leave comments if you want to address them yourself, or I can pull it into a branch and apply the fixes? Whichever you prefer.

**Comment by Lctrs** [comment]
Just leave me comments. Will adress them tomorrow at the latest. Thanks.

**Comment by Lctrs** [comment]
@frankdejonge I adressed what I saw. Tell me if you spotted more of them.

**Comment by frankdejonge** [comment]
@Lctrs can we rename the adapter namespace from `Zip` to `ZipArchive`? I'll go over the code now, didn't have time yesterday to review.

File: src/Zip/OpenZipException.php [hunk 7]
```diff
@@ -0,0 +1,11 @@
+<?php
+
+declare(strict_types=1);
+
+namespace League\Flysystem\Zip;
+
+use League\Flysystem\FilesystemException;
+
+interface OpenZipException extends FilesystemException
```
Comment (frankdejonge) [inline src/Zip/OpenZipException.php]: Please rename to: ZipArchiveException

File: src/Zip/ZipAdapter.php [hunk 8]
```diff
@@ -0,0 +1,369 @@
...
+        }
+
+        if ($this->archive()->deleteName($this->prefixer->prefixPath($path))) {
+            return;
+        }
+
+        throw UnableToDeleteFile::atLocation($path);
+    }
+
+    public function deleteDirectory(string $path): void
```
Comment (frankdejonge) [inline src/Zip/ZipAdapter.php]: Does this implementation also delete a directory with files in it? (it should)

File: src/Zip/composer.json [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+        "league/flysystem": "^2.0.0-RC1",
+        "league/mime-type-detection": "^1.0.0"
+    },
+    "license": "MIT",
+    "authors": [
+        {
+            "name": "Frank de Jonge",
+            "email": "info@frankdejonge.nl"
+        },
+        {
```
Comment (frankdejonge) [inline src/Zip/composer.json]: Although I'm a big fan of giving credit where credit is due, I've always restricted these sections for maintainers. Do you have a twitter handle we can reference in the docs instead? We can add a "contributed by" with a link to your profile.

File: src/Zip/ZipAdapter.php (hunk 8 above)
Comment (Lctrs) [reply src/Zip/ZipAdapter.php]: It does not. Oops.

File: src/Zip/OpenZipException.php (hunk 7 above)
Comment (Lctrs) [reply src/Zip/OpenZipException.php]: Adressed.

File: src/Zip/ZipAdapter.php (hunk 8 above)
Comment (Lctrs) [reply src/Zip/ZipAdapter.php]: Adressed.

File: src/Zip/composer.json (hunk 9 above)
Comment (Lctrs) [reply src/Zip/composer.json]: Adressed. Just link to my GitHub profile please.

**Comment by frankdejonge** [comment]
@Lctrs the tests are still failing, have you looked into that? Again, I'm more than willing to take it over at this point, so just let me know 👍

**Comment by Lctrs** [comment]
@frankdejonge Yes I'm looking into it. Path handling is a pain in the *ss, but I'll make it work.

**Comment by Lctrs** [comment]
@frankdejonge Failure unrelated. Should be good now.

File: src/ZipArchive/ZipArchiveAdapter.php [hunk 10]
```diff
@@ -0,0 +1,401 @@
...
+        if ($contents === false) {
+            throw UnableToReadFile::fromLocation($path, $this->archive()->getStatusString());
+        }
+
+        return $contents;
+    }
+
+    public function readStream(string $path)
+    {
+        $resource = $this->archive()->getStream($this->pathNormalizer->forFile($path));
```
Comment (Lctrs) [inline src/ZipArchive/ZipArchiveAdapter.php:117]: Be aware that this stream is not seekable. Don't know if that's a blocker or not.

**Comment by frankdejonge** [comment]
Thank you @Lctrs, I'll merge this in this week 👍

File: src/ZipArchive/ZipArchivePathNormalizerTest.php [hunk 11]
```diff
@@ -0,0 +1,154 @@
+<?php
+
+declare(strict_types=1);
+
+namespace League\Flysystem\ZipArchive;
+
+use PHPUnit\Framework\TestCase;
+
+final class ZipArchivePathNormalizerTest extends TestCase
```
Comment (Lctrs) [inline src/ZipArchive/ZipArchivePathNormalizerTest.php:9]: @frankdejonge I think this outline the need to standardize paths given by the filesystem to the adapter and the ones returned to the user by the filesystem.

**Pull request merged**